java-client/src/main/java/org/opensearch/client/transport/aws/AwsV2OpenSearchTransport.java

```diff
@@ -0,0 +1,427 @@
+        applyOptionsParams(url, options);
+        try {
+            req.uri(new URI(url.toString()));
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException("Invalid request URI: " + url.toString());
+        }
+        applyOptionsHeaders(req, transportOptions);
+        applyOptionsHeaders(req, options);
+        if (endpoint.hasRequestBody() && body != null) {
+            req.putHeader("Content-Type", "application/json");
+            req.putHeader("Content-Length", String.valueOf(body.length));
+            req.contentStreamProvider(() -> new ByteArrayInputStream(body));
+        }
+
+        Aws4SignerParams signerParams = Aws4SignerParams.builder()
+                .awsCredentials(credentials.resolveCredentials())
+                .signingName("es")
+                .signingRegion(Region.US_EAST_1)
+                .build();
+        return Aws4Signer.create().sign(req.build(), signerParams);
+    }
+
+    private void applyOptionsParams(StringBuilder url, TransportOptions options) {
+        if (options == null) {
+            return;
+        }
+        Map<String, String> params = options.queryParameters();
+        if (params != null && !params.isEmpty()) {
+            char sep = url.indexOf("?") < 0 ? '?' : '&';
+            for (Map.Entry<String, String> param : params.entrySet()) {
+                url.append(sep).append(param.getKey()).append('=');
+                url.append(URLEncoder.encode(param.getValue(), StandardCharsets.UTF_8));
+                sep = '?';
+            }
+        }
+    }
+
+    private void applyOptionsHeaders(SdkHttpFullRequest.Builder builder, TransportOptions options) {
+        if (options == null) {
+            return;
+        }
+        Collection<Map.Entry<String, String>> headers = options.headers();
+        if (headers != null && !headers.isEmpty()) {
+            for (Map.Entry<String, String> header : headers) {
+                builder.putHeader(header.getKey(), header.getValue());
+            }
+        }
+    }
+
+    private <ResponseT, ErrorT> ResponseT parseResponse(
+            SdkHttpResponse httpResponse,
+            InputStream bodyStream,
+            Endpoint<?, ResponseT, ErrorT> endpoint
+    ) throws IOException {
+        int statusCode = httpResponse.statusCode();
+        if (endpoint.isError(statusCode)) {
+            JsonpDeserializer<ErrorT> errorDeserializer = endpoint.errorDeserializer(statusCode);
+            if (errorDeserializer == null || bodyStream == null) {
+                throw new TransportException(
+                        "Request failed with status code '" + statusCode + "'"
+                );
+            }
+            try {
+                try (JsonParser parser = mapper.jsonProvider().createParser(bodyStream)) {
+                    ErrorT error = errorDeserializer.deserialize(parser, mapper);
+                    throw new OpenSearchException((ErrorResponse) error);
```

Comment:
Should we log the error here as well?

Reply:
I don't think so.  It'll get logged where it gets caught